---
title: "Merge Seurat Objects `r params$sample_name`"
author: 'User ID: `r Sys.getenv("USER")`'
date: "`r Sys.Date()`"
output:
  html_document:
    fig_height: 7
    fig_width: 9
    keep_md: yes
    toc: yes
    df_print: paged
params:
  seurat_obj_path:
    value:
      # count1.701: path to data
  sample_name: ""
  assay: "RNA"   # Assay we are working with
  grouping: "orig.ident" # grouping for plots
  hash_grouping: "hash.ID" # grouping for plots hashtags but doesnt work with anything else right now
  min_genes: 400 # minimum number of genes per cell
  max_genes: 6000 # maximum number of genes per cell 
  max_mt: 15  # maximum percent mitochondrial reads per cell 
  normalize_method: "log_norm" # normalization method. One of log_norm or sct
  nfeatures: 2000 # number of variable features to extract
  num_pcs: 30 # number of PCs to calculate
  num_dim: [20, 25] # number of PCs to use in UMAP
  num_neighbors: [20, 10] # number of neighbors to use in UMAP
  prefix: "lognorm"   # prefix to add to UMAP. Useful if you are doing different normalizations, or using different subsets of the data
  log_file: "log" # log file
  out_path: "."
---

```{r setup, include=FALSE}
attach(params) 
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE, 
                      message = FALSE, 
                      fig.path = file.path(out_path, paste0("Merge-", sample_name, "/")),
                      fig.height = 10, 
                      fig.width = 15,
                      dev = c("png", "pdf"))
```

```{r load_libraries}
library(future)
library(cowplot)
library(devtools)
library(forcats)
library(GGally)
library(stringr)
library(ggpmisc)
library(ggrepel)
load_all("/gpfs/home/ay1392/tsirigoslabdir/scooter") # PATH TO SCOOTER
theme_set(theme_bw())

# # evaluate Seurat R expressions asynchronously when possible (such as ScaleData) using future package
plan("multiprocess", workers = 4)
# # increase the limit of the data to be shuttled between the processes from default 500MB to 50GB
options(future.globals.maxSize = 50 * 1024 ^ 3)

```

```{r process_params}
# Create a new directory for this sample. Error is that directory already exists
out_path = file.path(out_path, paste0("Merge-", sample_name, "/"))

if (dir.exists(out_path)) {
  stop(glue("output analysis dir {out_path} already exists"))
} else {
  dir.create(out_path)
}

yaml::write_yaml(params, glue("{out_path}/params.yml"))
log_file = file.path(out_path, "log")
```

## Create RMD

The output of this file is a directory called `r paste0("Merge-", sample_name)` within which there will be a params.yml that saves the parameters from this file, and a log file. 

If you choose to use the cache = TRUE parameter, rendering this file may be quicker the second time around, but please be aware that when you run the file a second time, the cached code chunk will not run again unless you change something *within* the code chunk. So, if you make changes *upstream* of the cached code chunk, the cached code chunk will not take that into account and output the cached result. 

# Read in seurat objects
`r seurat_obj_path`

```{r read_in}
seurat_obj_list <- gather_seurat_objects(seurat_obj_path, assay)
```

# Merge the libraries
```{r merge_libraries}
seurat_obj_list <- seurat_obj_list[[1]]
seurat_obj = Seurat:::merge.Seurat(seurat_obj_list[[1]], seurat_obj_list[2:length(seurat_obj_list)])
rm(seurat_obj_list)
```


Plot distribution of the number of genes, UMI, and percent mitochondrial reads per cell. 
```{r unfiltered_violin_qc}
# plot number of genes 
genes_unfilt <- plot_distribution(seurat_obj, 
                                features = "nFeature_RNA", 
                                grouping = grouping) +
  geom_hline(yintercept=min_genes, color = "red", size = 2) +
  geom_hline(yintercept=max_genes, color = "red", size = 2) +
  annotate(geom="text", x=0.7, y=max_genes - 200, label= glue("Max genes: {max_genes}"),
              color="red") +
  annotate(geom="text", x=0.7, y=min_genes - 200, label=glue("Min genes: {min_genes}"),
              color="red")


# plot number of UMIs
umi_unfilt <- plot_distribution(seurat_obj, 
                              features = "nCount_RNA",
                              grouping = grouping) 

# plot percent mitochondrial reads
mito_unfilt <- plot_distribution(seurat_obj, 
                               features = "pct_mito",
                               grouping = grouping) +
  geom_hline(yintercept=max_mt, color = "red", size = 2) +
  annotate(geom="text", x=0.7, y=max_mt + 5, label=glue("Max pct mito: {max_mt}"),
              color="red")

# get the legend for one of the plots to use as legend for the combined plot
legend_grid <- get_legend(mito_unfilt)

title <- ggdraw() + 
  draw_label(
    "Unfiltered-data",
    fontface = 'bold',
    x = 0,
    hjust = 0
  ) +
  theme(
    # add margin on the left of the drawing canvas,
    # so title is aligned with left edge of first plot
    plot.margin = margin(0, 0, 0, 7)
  )

# Combine plots 
plot_row <- plot_grid(genes_unfilt + theme(legend.position = "none"),
              umi_unfilt + theme(legend.position = "none"),
              mito_unfilt + theme(legend.position = "none"),
              legend_grid,
              ncol = 4)

plot_grid(title,
          plot_row,
          ncol = 1,
          rel_heights = c(0.1, 1))
```


Filter the data.
```{r filter}
# filter data
seurat_obj <- filter_data(seurat_obj, min_genes = min_genes,
                          max_genes = max_genes, max_mt = max_mt,
                          log_file = log_file)
```

```{r freq_mean}
# for each gene, calculate % of cells expressing it and the mean
pct_cells <- apply(as.data.frame(seurat_obj@assays$RNA@counts), 1, function(x) (sum(x != 0))) / ncol(seurat_obj@assays$RNA@counts)

gene_means <- rowMeans(as.data.frame(seurat_obj@assays$RNA@counts))

p <- as.data.frame(cbind(pct_cells, gene_means)) %>% 
  rownames_to_column("gene")
ggplot(data = p, aes(x = gene_means, y = pct_cells, label = gene)) +
  geom_point(size = 3, alpha = 0.6) +
  xlab("Mean count per gene") +
  ylab("Percent cells expressing gene") +
  stat_dens2d_labels(geom = "text_repel", keep.fraction = 0.001)
```

```{r cell-activity}
# number of expressed genes
num_genes <- apply(as.data.frame(seurat_obj@assays$RNA@counts), 2, function(x) (sum(x != 0)))

libsize <- colSums(as.data.frame(seurat_obj@assays$RNA@counts))

l  <- as.data.frame(cbind(num_genes, libsize)) 

ggplot(data = l, aes(x = log10(libsize), y = num_genes)) +
  geom_point(size = 3, alpha = 0.6) +
  xlab("Library Size (log10)") +
  ylab("Number of Expressed Genes") 
```

```{r filtered_qc}
# plot number of genes
genes_sing <- plot_distribution(seurat_obj,
                                features = "nFeature_RNA",
                                grouping = grouping) +
  geom_hline(yintercept=min_genes, color = "red", size = 2) +
  geom_hline(yintercept=max_genes, color = "red", size = 2) +
  annotate(geom="text", x=0.7, y=max_genes - 200, label= glue("Max genes: {max_genes}"),
              color="red") +
  annotate(geom="text", x=0.7, y=min_genes - 200, label=glue("Min genes: {min_genes}"),
              color="red")


# plot number of UMIs
umi_sing <- plot_distribution(seurat_obj,
                              features = "nCount_RNA",
                              grouping = grouping)

# plot percent mitochondrial reads
mito_sing <- plot_distribution(seurat_obj,
                               features = "pct_mito",
                               grouping = grouping)+
    geom_hline(yintercept=max_mt, color = "red", size = 2) +
  annotate(geom="text", x=0.7, y=max_mt + 5, label=glue("Max pct mito: {max_mt}"),
              color="red")

# get the legend for one of the plots to use as legend for the combined plot
legend_grid <- get_legend(mito_sing)

# Combine plots
plot_grid(genes_sing + theme(legend.position = "none"),
          umi_sing + theme(legend.position = "none"),
          mito_sing + theme(legend.position = "none"),
          legend_grid,
          ncol = 4)
```

```{r filtered_hash_qc}

genes_sing <- plot_distribution(seurat_obj,
                                features = "nFeature_RNA",
                                grouping = hash_grouping) +
geom_hline(yintercept=min_genes, color = "red", size = 2) +
geom_hline(yintercept=max_genes, color = "red", size = 2) +
annotate(geom="text", x=0.7, y=max_genes - 200, label= glue("Max genes: {max_genes}"),
            color="red") +
annotate(geom="text", x=0.7, y=min_genes - 200, label=glue("Min genes: {min_genes}"),
            color="red")

# plot number of UMIs
umi_sing <- plot_distribution(seurat_obj,
                              features = "nCount_RNA",
                              grouping = hash_grouping)

# plot percent mitochondrial reads
mito_sing <- plot_distribution(seurat_obj,
                               features = "pct_mito",
                               grouping = hash_grouping) +
  geom_hline(yintercept=max_mt, color = "red", size = 2) +
annotate(geom="text", x=0.7, y=max_mt + 5, label=glue("Max pct mito: {max_mt}"),
            color="red")

# get the legend for one of the plots to use as legend for the combined plot
legend_grid <- get_legend(mito_sing)

# Combine plots
plot_row <- plot_grid(genes_sing + theme(legend.position = "none"),
          umi_sing + theme(legend.position = "none"),
          mito_sing + theme(legend.position = "none"),
          legend_grid,
          ncol = 4)

title <- ggdraw() + 
draw_label(
  "Unfiltered-data",
  fontface = 'bold',
  x = 0,
  hjust = 0
) +
theme(
  # add margin on the left of the drawing canvas,
  # so title is aligned with left edge of first plot
  plot.margin = margin(0, 0, 0, 7)
)

plot_grid(title,
          plot_row,
          ncol = 1,
          rel_heights = c(0.1, 1))

```


```{r adt_freq_mean}
# for each gene, calculate % of cells expressing it and the mean
pct_cells <- apply(as.data.frame(seurat_obj@assays$ADT@data), 1, function(x) (sum(x != 0))) / ncol(seurat_obj@assays$ADT@data)

gene_means <- rowMeans(as.data.frame(seurat_obj@assays$ADT@data))

p <- as.data.frame(cbind(pct_cells, gene_means)) %>% 
  rownames_to_column("ADT")
ggplot(data = p, aes(x = gene_means, y = pct_cells, label = ADT)) +
  geom_point(size = 3, alpha = 0.6) +
  xlab("Mean count per ADT") +
  ylab("Percent cells expressing ADT") +
  stat_dens2d_labels(geom = "text_repel", keep.fraction = 0.001)

```

Normalize the data using `r normalize_method`
```{r Normalize}
 # normalize seurat object using specified method, on specified assay
 seurat_obj <- normalize_data(seurat_obj, method = normalize_method,
                              nfeatures = nfeatures, assay = assay)
```

```{r runpca}
# run PCA on specified assay using the number of PCs specified
seurat_obj <- run_dr(data = seurat_obj,
                  dr_method = "pca",
                  var_features = TRUE,
                  assay = assay,
                  num_pcs = num_pcs,
                  prefix =  prefix)
```

```{r dimred}
# Create a dataframe of all of the possible combinations of number of PCs to use for
# UMAP, and the number of neighbors
num_dim_vect <- c(num_dim)
num_neighbors_vect <- c(num_neighbors)
possibilities <- expand.grid(num_dim_vect, num_neighbors_vect)

# For each of these combinations, calculate UMAP
for(i in 1:nrow(possibilities)) {
  num_dim <- possibilities[i, 1]
  num_neighbors <- possibilities[i, 2]
  seurat_obj <- run_dr(data = seurat_obj, dr_method = "umap", reduction = paste0("pca", prefix),
                     num_dim_use = num_dim, assay = "RNA", num_neighbors = num_neighbors,
                     prefix = glue("ndim{num_dim}nn{num_neighbors}{prefix}"))
}
```

Save metadata as a csv to `r glue("{out_path}/metadata_create_{sample_name}.csv")`
```{r save_metadata}
# Export the reductions to Seurat and save dimensionality reduction and
# metadata as a csv
reduction_names <- c(paste0("umap", "ndim", possibilities[,1], "nn", possibilities[,2], prefix), paste0("pca", prefix))

metadata <- as_data_frame_seurat(seurat_obj, reduction = reduction_names,
                                 metadata = TRUE)

write_excel_csv(metadata, path = glue("{out_path}/metadata_create_{sample_name}.csv"))
```

Plot dimensionality reduction.
```{r plotdr}
# plot the first two PCs, and all of the different UMAPs
reduction_names <- c(paste0("UMAP", "ndim", possibilities[,1], "nn", possibilities[,2], prefix), paste0("PC", prefix))

plot_dr <- data.frame(X = paste0(reduction_names, "_1"),
                      Y = paste0(reduction_names, "_2"),
                      stringsAsFactors = FALSE)

dir.create(file.path(out_path, "dr"), showWarnings = FALSE)

for(i in 1:nrow(plot_dr)){
  print(current_plot <- plot_scatter(metadata = metadata,
                               out_path = file.path(out_path, "dr"),
                               proj_name = sample_name,
                               log_file = log_file,
                               X = plot_dr[i,1],
                               Y = plot_dr[i,2],
                               color = grouping,
                               write = TRUE))
  
    print(current_plot <- plot_scatter(metadata = metadata,
                               out_path = file.path(out_path, "dr"),
                               proj_name = sample_name,
                               log_file = log_file,
                               X = plot_dr[i,1],
                               Y = plot_dr[i,2],
                               color = "nFeature_RNA",
                               write = TRUE))
    
    print(current_plot <- plot_scatter(metadata = metadata,
                           out_path = file.path(out_path, "dr"),
                           proj_name = sample_name,
                           log_file = log_file,
                           X = plot_dr[i,1],
                           Y = plot_dr[i,2],
                           color = "nCount_RNA",
                           write = TRUE))
    
    print(current_plot <- plot_scatter(metadata = metadata,
                       out_path = file.path(out_path, "dr"),
                       proj_name = sample_name,
                       log_file = log_file,
                       X = plot_dr[i,1],
                       Y = plot_dr[i,2],
                       color = "pct_mito",
                       write = TRUE))
}
```

Save rds to `r paste0(out_path, "/", "seurat_obj.rds")`
```{r save_rds}
# save final Seurat object
saveRDS(seurat_obj, file = paste0(out_path, "/", "seurat_obj.rds"))
```

