---
title: "Differential Expression for set4 set5 cluster resolution 0.9"
author: 'User ID: `r Sys.getenv("USER")`'
date: "`r Sys.Date()`"
output:
  html_document:
    fig_height: 7
    fig_width: 9
    keep_md: yes
    toc: yes
    df_print: paged
params:
  metadata_path: "/home/ay1392/anna_beegfs/TET2/scRNAseq/Analysis/WBM/Cluster-set4-set5/cluster-dim.50-neighbors.30/metadata.csv" # metadata with the clusters/keys
  data_path: "" # directory with the seurat object
  out_path: "."
  sample_name: ""
  keys: "res.0.9" # metadata column to group cells with 
  list_groups: NULL
---

```{r setup, include=FALSE}
attach(params) 
library(devtools)
load_all("/home/ay1392/anna_beegfs/TET2/scRNAseq/scooter")
theme_set(theme_bw())
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      fig.path = file.path(out_path, paste0("DE-", sample_name, "/")))
```

```{r process_params}
out_path = file.path(out_path, paste0("DE-", sample_name))
dir.create(out_path, showWarnings = FALSE)
```

# Read in data from somewhere a s_obj 
```{r}
seurat_obj <- readRDS(paste0(data_path, "/seurat_obj.rds"))
data <- seurat_obj@assays$RNA@data

metadata <- read_csv(metadata_path)
metadata_column <- keys
```

```{r}
DE_global <- differential_expression_global(data = data, metadata = metadata, 
                                            metadata_column = metadata_column,
                                            log_fc_threshold = 0.5, 
                                            min.pct = 0.5, test.use = "wilcox",
                                            out_path = out_path, write = TRUE,
                                            log_file = NULL)
```

```{r}
saveRDS(DE_global, file = paste0(out_path, "/DE_global.rds"))
```

```{r}
DE_paired <- differential_expression_paired(data = data, metadata = metadata, 
                                     metadata_column = metadata_column,
                                     list_groups = list_groups, log_fc_threshold = 0.5, 
                                     min.pct = 0.5, test.use = "wilcox",
                                     out_path = out_path, write = FALSE, 
                                     log_file = NULL)
```

```{r}
saveRDS(DE_paired, file = paste0(out_path, "/DE_paired.rds"))
```


<!-- ```{r} -->
<!-- visualize_top_de <- function(de_df, data, logfc_cutoff, adj_pval_cutoff, out_path) { -->
<!--   # get genes that pass cutoff -->
<!--   #heatmap of everything -->

<!--   de_df <- de_df %>% -->
<!--     filter(avg_logFC > logfc_cutoff | avg_logFC < -logfc_cutoff) %>%  -->
<!--     filter(p_val_adj < adj_pval_cutoff) -->

<!--   # get these genes from data -->
<!--   data_heat <- data[which(rownames(data) %in% de_df$gene),] -->

<!--   pdf(glue("{out_path}heatmap-{de_df$group_1}-{de_df$group_2}.pdf")) -->
<!--   pheatmap(data_heat) -->
<!--   dev.off() -->

<!--   # get top 6 on each side -->
<!--   # umaps -->

<!-- } -->
<!-- ``` -->



<!-- ```{r} -->
<!-- mainDir = "." -->

<!-- metadata <- read_csv(metadata_path) -->
<!-- metadata_column <- keys -->

<!-- seurat_obj <- readRDS(paste0(data_path, "/seurat_obj.rds")) -->
<!-- data <- seurat_obj@assays$RNA@data -->
<!-- seurat_obj@meta.data <- seurat_obj@meta.data %>%  -->
<!--     column_to_rownames("cell") -->

<!-- # create a directory for each cluster -->
<!-- for(i in 1:length(DE_global)){ -->
<!--   cluster <- strsplit(names(DE_global)[i], ".All")[[1]][1] -->
<!--   dir.create(file.path(mainDir, cluster), showWarnings = FALSE) -->
<!--   out_dir = paste0(mainDir, "/",cluster , "/" ) -->

<!--   colors <- rep("grey", nrow(unique(metadata[keys]))) -->
<!--   names(colors) <- unique(metadata[keys])[[1]] -->
<!--   colors[cluster] <- "black" -->

<!--   umap <- ggplot(metadata, aes(UMAP30lognorm_1, UMAP30lognorm_2, color = as.character(!!sym(keys)))) + -->
<!--     geom_point(alpha = 0.7 ) +  -->
<!--     theme_bw() + -->
<!--     scale_color_manual(values = colors) + -->
<!--     xlab("UMAP 1") + -->
<!--     ylab("UMAP 2") -->

<!--   ggsave(umap, file = paste0(out_dir, cluster, ".png")) -->

<!--   current_DE <- DE_global[[i]] -->
<!--   if(nrow(current_DE) > 5) { -->
<!--     #visualize_top_de(current_DE, data, logfc_cutoff = 0, adj_pval_cutoff= 0.05, out_path = out_dir) -->
<!--   } -->

<!--   current_DE <- current_DE[order(current_DE$avg_logFC, decreasing = TRUE),] -->
<!--   iter <- ifelse(nrow(current_DE) > 5, 5, nrow(current_DE)) -->

<!--   umap_genes <- current_DE$gene[1:iter] -->

<!--   current_metadata <- as_data_frame_seurat(seurat_obj, metadata = TRUE, reduction = "umap30lognorm", assay = "RNA", slot = "data", features = umap_genes) -->

<!--   for(j in 1:iter){ -->
<!--     current_gene <- umap_genes[j] -->
<!--     umap <- ggplot(current_metadata, aes(UMAP30lognorm_1, UMAP30lognorm_2, color = !!sym(current_gene))) + -->
<!--       geom_point(alpha = 0.7) +  -->
<!--       theme_bw() + -->
<!--       xlab("UMAP 1") + -->
<!--       ylab("UMAP 2") + -->
<!--       ggsci::scale_color_material("red") -->

<!--     ggsave(umap, file = paste0(out_dir, cluster, current_gene, ".png")) -->

<!--   } -->
<!-- } -->

<!-- # plot UMAp with the current cluster -->

<!-- # plot heatmap of most diff exp genes -->

<!-- # plot umap of the top 5 diff exped genes -->
<!-- ``` -->


